### DISCUSSION TITLE
How to pass inputs from agents/Tasks to a custom tool in crew ai

### DISCUSSION THREAD

**Discussion by pradeepdev-1995** [crewAIInc/crewAI#1096] [Q&A]
How to pass inputs from agents to a custom tool in crew ai . Here is my code snippet
```
import os
from crewai import Agent, Task, Crew, Process

from crewai_tools import BaseTool

import openai
import os
from langchain.chat_models.azure_openai import AzureChatOpenAI
agentLlm = AzureChatOpenAI(<my credentials>)
class FileWriterTool(BaseTool):
    name: str = "FileWriter"
    description: str = "Writes given content to a specified file."

    def _run(self, filename: str, content: str) -> str:
        # Open the specified file in write mode and write the content
        with open(filename, 'w') as file:
            file.write(content)
        return f"Content successfully written to {filename}"

# Set up the FileWriterTool
file_writer = FileWriterTool()

# Define the agent with a role, goal, and tools
researcher = Agent(
    role='Knowledge Article Writer',
    goal='Create and save detailed content on professional domains to a file.',
    backstory="Passionate about crafting in-depth articles",
    verbose=True,
    allow_delegation=False,
    llm=agentLlm,
    tools=[file_writer]
)

# Create a task that utilizes the FileWriterTool to save content
task1 = Task(
    description="Write and save an article using the FileWriter tool.",
    expected_output="A file with the article content more than 1000 words.",
    agent=researcher,
    tools=[file_writer],
    function_args={'filename': 'deep_learning.md', 'content': 'Detailed content generated by LLM about deep_learning.'}
)

# Instantiate the crew with a sequential process and execute the task
crew = Crew(
    agents=[researcher],
    tasks=[task1],
    process=Process.sequential,
    verbose=2
)

# Execute the crew tasks and print the result
result = crew.kickoff()
print("######################")
print(result)
```
What I observe is, the parameters in ```function_args={'filename': 'deep_learning.md', 'content': 'Detailed content generated by LLM about deep_learning.'}``` are not reaching in tools and it randomly taking the topics from LLM.

**Comment by shnoman97** [comment]
@pradeepdev-1995  First, there is an arg in Agents to output the final response in a file. And for passing the file name, you need to prompt the name of the file into agent via goal.

**Comment by pradeepdev-1995** [comment]
@shnoman97 
If i should mention the topic/file name in the prompt, then whatbis the role of func_args in  the task.
Iam expect that the my arguments will reach to the custom tool dynamically, without a hardcoded prompt.

**Comment by shnoman97** [comment]
There is no `function_args ` opt available in Tasks.
https://docs.crewai.com/core-concepts/Tasks/#task-attributes

**Comment by pradeepdev-1995** [comment]
@shnoman97 
Okay then i how i can execute the custom tool dynamically based on the arguments in each crew kickoff.

I dont want to hardcode the details in the
Goal/expected output prompt for tasks/agents.

Because in every kick of the crew i want to pass different arguments to the custom tool.

**Comment by shnoman97** [comment]
You should learn about `Prompt Templating`.
See: https://docs.crewai.com/core-concepts/Agents/#creating-an-agent

**Reply by mayonesa** [reply]
Prompt templating is [here](https://docs.crewai.com/concepts/agents#setting-prompt-templates) instead.
